### ISSUE TITLE
Accelerate copy_local_to_global

### ISSUE THREAD

**Issue by bangerth** [geodynamics/aspect#756]
I don't know if this is worth it or not, but here's an idea: For example, for the temperature system, we have

```
  template <int dim>
  void
  Simulator<dim>::
  copy_local_to_global_advection_system (const internal::Assembly::CopyData::AdvectionSystem<dim> &data)
  {
    // copy entries into the global matrix. note that these local contributions
    // only correspond to the advection dofs, as assembled above
    current_constraints.distribute_local_to_global (data.local_matrix,
                                                    data.local_rhs,
                                                    data.local_dof_indices,
                                                    system_matrix,
                                                    system_rhs);
  }
```

but we know, in reality, that `data.local_matrix` and `data.local_rhs` only set entries that correspond to the current set of advection DoFs. In other words, `distribute_local_to_global()` will do _a lot_ of work that is not actually necessary, by distributing zeros.

It would be interesting to time how much this function actually costs, and maybe provide a mask to this function regarding the vector and matrix entries it should care about.

**Comment by bangerth** [comment]
(Alternatively, set the DoF indices of DoFs we don't care about to `invalid_dof_index` or similar.)

**Comment by gassmoeller** [comment]
At the same time when changing the `copy_local_to_global_..._system` function to only distribute the relevant dofs, one might also look into the `local_assemble_stokes_system/preconditioner` functions, and let them only loop over relevant dofs as well, as discussed in #1246. This is mostly a note to not forget about that.

**Comment by tjhei** [comment]
The way I would go about it is something like an ``FEValuesSubset`` that takes an FEValues and a bitfield of components to select and then returns a reduced ``n_dofs`` and ``get_dof_indices``.

**Comment by class4kayaker** [comment]
Looking through the code, I believe that for the advection assembly, the list of DoFs is already restricted to the FE for the current component. As such, only the stokes assembly does the assembly for all of the DoFs. I might be overlooking something however.

**Comment by gassmoeller** [comment]
@class4kayaker is right. Although it is very misleading, everything looks similar for the advection and stokes case, but the local advection matrix is sized for the `advection_element.dofs_per_cell`, while the stokes case is sized for `finite_element.dofs_per_cell`. So only the Stokes case needs a change.

**Comment by bangerth** [comment]
I've changed my mind thinking that one should add a parameter to `distribute_local_to_global`. We should just shrink the matrix to that sub-matrix that is actually non-zero. If we shrink the vector of global dof indices the same way, then we can continue to call `distribute_local_to_global` -- it will simply not see the zero rows and columns, and corresponding global dof indices. 

We could do this shrinking in `copy_local_to_global`, or already during assembly. (This has the advantage that we don't need to be inconvenienced by having to change deal.II.)

**Comment by gassmoeller** [comment]
This was solved by #1250.
